src/state/apis/foxy/foxyBalancesApi.ts

```diff
@@ -0,0 +1,227 @@
+import {
+  selectAssets,
+  selectMarketData,
+  selectPortfolioAssetBalances,
+  selectPortfolioLoading,
+} from 'state/slices/selectors'
+
+import { getFoxyApi } from './foxyApiSingleton'
+
+type GetFoxyBalancesInput = {
+  userAddress: string
+  foxyApr: string
+}
+type GetFoxyBalancesOutput = {
+  opportunities: MergedFoxyOpportunity[]
+  totalBalance: string
+}
+
+export type FoxyOpportunity = {
+  type: DefiType
+  provider: string
+  version: string
+  contractAddress: string
+  rewardToken: string
+  stakingToken: string
+  chainId: ChainId
+  tvl?: string
+  expired?: boolean
+  apy?: string
+  balance: string
+  contractAssetId: AssetId
+  tokenAssetId: AssetId
+  rewardTokenAssetId: AssetId
+  pricePerShare: string
+  withdrawInfo: WithdrawInfo
+}
+
+export type MergedFoxyOpportunity = FoxyOpportunity & {
+  cryptoAmount: string
+  fiatAmount: string
+}
+
+const moduleLogger = logger.child({
+  namespace: ['state', 'apis', 'foxy', 'UseFoxyBalances'],
+})
+
+type MaybeMarketCapData = Record<AssetId, MarketData | undefined>
+
+const makeFiatAmount = (
+  opportunity: FoxyOpportunity,
+  assets: AssetsById,
+  marketData: MaybeMarketCapData,
+) => {
+  const asset = assets[opportunity.tokenAssetId]
+  const pricePerShare = bnOrZero(opportunity.pricePerShare).div(`1e+${asset?.precision}`)
+  const marketPrice = marketData[opportunity.tokenAssetId]?.price
+  return bnOrZero(opportunity.balance)
+    .div(`1e+${asset?.precision}`)
+    .times(pricePerShare)
+    .times(bnOrZero(marketPrice))
+}
+
+const makeTotalBalance = (
+  opportunities: Record<string, FoxyOpportunity>,
+  assets: AssetsById,
+  marketData: MaybeMarketCapData,
+): BN =>
+  Object.values(opportunities).reduce((acc: BigNumber, opportunity: FoxyOpportunity) => {
+    const amount = makeFiatAmount(opportunity, assets, marketData)
+    return acc.plus(bnOrZero(amount))
+  }, bn(0))
+
+const makeMergedOpportunities = (
+  opportunities: Record<string, FoxyOpportunity>,
+  assets: AssetsById,
+  marketData: MaybeMarketCapData,
+) =>
+  Object.values(opportunities).map(opportunity => {
+    const asset = assets[opportunity.tokenAssetId]
+    const fiatAmount = makeFiatAmount(opportunity, assets, marketData)
+    const marketPrice = marketData[opportunity.tokenAssetId]?.price ?? 0
+    const tvl = bnOrZero(opportunity.tvl)
+      .div(`1e+${asset?.precision}`)
+      .times(marketPrice)
+      .toString()
+    const data = {
+      ...opportunity,
+      tvl,
+      cryptoAmount: bnOrZero(opportunity.balance).div(`1e+${asset?.precision}`).toString(),
+      fiatAmount: fiatAmount.toString(),
+    }
+    return data
+  })
+
+async function getFoxyOpportunities(
+  balances: PortfolioBalancesById,
+  api: FoxyApi,
+  userAddress: string,
+  foxyApr: string,
+) {
+  const acc: Record<string, FoxyOpportunity> = {}
+  try {
+    const opportunities = await api.getFoxyOpportunities()
+    for (let index = 0; index < opportunities.length; index++) {
```

Comment:
We could make this less mutatey:

```suggestion
    for (const [index] of opportunities.entries()) {
```

Reply:
Added to the checklist, will refactor in a follow-up stacked PR to keep things atomic so this can just act as a "shuffled things around" PR